sdk/go/pulumi/types.go

```diff
@@ -238,41 +238,49 @@ func (o *OutputState) reject(err error) {
 	o.fulfill(nil, true, false, nil, err)
 }
 
+// awaitOnce is a single iteration of the "await" loop, using the condition variable as a lock to
+// guard accessing the fields to avoid tearing reads and writes.
+func (o *OutputState) awaitOnce(ctx context.Context) (interface{}, bool, bool, []Resource, error) {
+	if o == nil {
+		// If the state is nil, treat its value as resolved and unknown.
+		return nil, false, false, nil, nil
+	}
+
+	o.cond.L.Lock()
+	for o.state == outputPending {
+		if ctx.Err() != nil {
+			return nil, true, false, nil, ctx.Err()
+		}
+		o.cond.Wait()
+	}
+	defer o.cond.L.Unlock()
+
+	return o.value, o.known, o.secret, o.deps, o.err
+}
+
 func (o *OutputState) await(ctx context.Context) (interface{}, bool, bool, []Resource, error) {
 	known := true
 	secret := false
 	var deps []Resource
 
 	for {
-		if o == nil {
-			// If the state is nil, treat its value as resolved and unknown.
-			return nil, false, false, nil, nil
-		}
-
-		o.cond.L.Lock()
-		for o.state == outputPending {
-			if ctx.Err() != nil {
-				return nil, true, false, nil, ctx.Err()
-			}
-			o.cond.Wait()
-		}
-		o.cond.L.Unlock()
```

Comment:
So in fact this wouldn't unlock if ctx.Err() != nil branch returned first? Sounds like we always want to unlock. Good!